File: app/src/main/java/com/xomena/cmpfutboltfe/SelectPlaceActivity.java
package com.xomena.cmpfutboltfe;

import android.app.Dialog;
import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.content.SharedPreferences;
import android.graphics.drawable.Drawable;
import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.widget.AdapterView;
import android.widget.AutoCompleteTextView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.appcompat.app.ActionBar;
import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import androidx.core.content.res.ResourcesCompat;
import androidx.fragment.app.DialogFragment;
import androidx.fragment.app.FragmentManager;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.google.android.gms.common.ConnectionResult;
import com.google.android.gms.common.api.GoogleApiClient;
import com.google.android.gms.common.api.ResultCallback;
import com.google.android.gms.location.places.AutocompletePrediction;
import com.google.android.gms.location.places.Place;
import com.google.android.gms.location.places.PlaceBuffer;
import com.google.android.gms.location.places.Places;

import java.util.LinkedHashSet;
import java.util.Set;

public class SelectPlaceActivity extends AppCompatActivity
        implements GoogleApiClient.OnConnectionFailedListener,
            SelectPlacesAdapter.OnItemClickListener {

    protected GoogleApiClient mGoogleApiClient;
    private PlacesAutocompleteAdapter mAdapter;
    private SelectPlacesAdapter adapter;
    private static SelectPlaceActivity self;
    private final static String LOG_TAG = "SelectPlaceActivity";
    protected static final String STORED_KEYS = "PLACE_IDS";

    private AdapterView.OnItemClickListener mAutocompleteClickListener
            = new AdapterView.OnItemClickListener() {
        @Override
        public void onItemClick(AdapterView<?> parent, View view, int position, long id) {
            /*
             Retrieve the place ID of the selected item from the Adapter.
             The adapter stores each Place suggestion in a AutocompletePrediction from which we
             read the place ID and title.
              */
            final AutocompletePrediction item = mAdapter.getItem(position);
            final String placeId = item.getPlaceId();
            final CharSequence primaryText = item.getPrimaryText(null);

            FragmentManager fm = getSupportFragmentManager();
            SavePlaceDialog alertDialog = SavePlaceDialog.newInstance(getString(R.string.title_activity_select_place),
                    primaryText.toString(), placeId);
            alertDialog.show(fm, "fragment_alert");
        }
    };

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        self = this;

        mGoogleApiClient = new GoogleApiClient.Builder(this)
                .enableAutoManage(this, 0 /* clientId */, this)
                .addApi(Places.GEO_DATA_API)
                .build();

        setContentView(R.layout.activity_select_place);
        Toolbar toolbar = (Toolbar) findViewById(R.id.toolbarSelectPlace);
        setSupportActionBar(toolbar);
        try {
            ActionBar ab = getSupportActionBar();
            if (ab != null) {
                ab.setDisplayShowTitleEnabled(false);
            }
        } catch (NullPointerException e) {
            Log.e(LOG_TAG, "Exception", e);
        }

        final Drawable upArrow = ResourcesCompat.getDrawable(getResources(), R.drawable.arrow_left,
                getApplicationContext().getTheme());
        toolbar.setNavigationIcon(upArrow);
        toolbar.setNavigationOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                onBackPressed();
            }
        });

        AutoCompleteTextView mAutocompleteView = (AutoCompleteTextView) findViewById(R.id.fieldFindPlace);
        mAutocompleteView.setOnItemClickListener(mAutocompleteClickListener);
        mAdapter = new PlacesAutocompleteAdapter(this, mGoogleApiClient, MainActivity.BOUNDS_TENERIFE, null);
        mAutocompleteView.setAdapter(mAdapter);

        RecyclerView recyclerView = (RecyclerView) findViewById(R.id.rvSelectPlace);

        adapter = new SelectPlacesAdapter(MyPlace.createPlacesList(this));
        adapter.setOnItemClickListener(this);

        recyclerView.setAdapter(adapter);
        // Set layout manager to position the items
        recyclerView.setLayoutManager(new LinearLayoutManager(this));
        RecyclerView.ItemDecoration itemDecoration = new
                DividerItemDecoration(this, DividerItemDecoration.VERTICAL_LIST);
        recyclerView.addItemDecoration(itemDecoration);
    }

    @Override
    public void onConnectionFailed(@NonNull ConnectionResult connectionResult) {

        Log.e(LOG_TAG, "onConnectionFailed: ConnectionResult.getErrorCode() = "
                + connectionResult.getErrorCode());

        Toast.makeText(this,
                "Could not connect to Google API Client: Error " + connectionResult.getErrorCode(),
                Toast.LENGTH_SHORT).show();
    }



    @Override
    public void onItemClick(View itemView, int position) {
        MyPlace place = adapter.getItem(position);
        if (place != null) {
            returnPlace(place.getPlaceID(), place.getAddress());
        } else {
            Intent resultIntent = new Intent();
            setResult(RESULT_CANCELED, resultIntent);
            finish();
        }
    }

    private void savePlace (String placeId, String placeName) {
        final SharedPreferences sharedPref = this.getSharedPreferences(
                MainActivity.PLACES_SHARED_PREF, Context.MODE_PRIVATE);

        final Set<String> keys = sharedPref.getStringSet(STORED_KEYS, new LinkedHashSet<String>());
        if (keys.contains(placeId)) {
            String toastMsg = String.format(getString(R.string.place_alredy_in_list), placeName);
            Toast.makeText(this, toastMsg, Toast.LENGTH_LONG).show();
        } else {
            Places.GeoDataApi.getPlaceById(mGoogleApiClient, placeId).setResultCallback(new ResultCallback<PlaceBuffer>() {
                @Override
                public void onResult(@NonNull PlaceBuffer places) {
                    if (places.getStatus().isSuccess()) {
                        Place place = places.get(0);

                        Set<String> placeData = new LinkedHashSet<>();
                        String val = place.getId() + "###" + String.valueOf(place.getLatLng().latitude) +
                                "###" + String.valueOf(place.getLatLng().longitude) +
                                "###" + place.getName().toString() +
                                "###" + place.getAddress().toString();
                        placeData.add(val);

                        keys.add(place.getId());

                        SharedPreferences.Editor editor = sharedPref.edit();
                        editor.putStringSet(STORED_KEYS, keys);
                        editor.putStringSet(place.getId(), placeData);
                        editor.apply();

                        String toastMsg = String.format(getString(R.string.place_added_to_list), place.getName());
                        Toast.makeText(self, toastMsg, Toast.LENGTH_LONG).show();
                    }
                    //Release the PlaceBuffer to prevent a memory leak
                    places.release();
                }
            });
        }
    }

    private void returnPlace (String placeId, String placeAddress) {
        Intent resultIntent = new Intent();
        resultIntent.putExtra(MainActivity.EXTRA_PLACEID, placeId);
        resultIntent.putExtra(MainActivity.EXTRA_ADDRESS, placeAddress);
        setResult(RESULT_OK, resultIntent);
        finish();
    }

    public static class SavePlaceDialog extends DialogFragment {
        public SavePlaceDialog() {
            // Empty constructor required for DialogFragment
        }

        public static SavePlaceDialog newInstance(String title, String placeName, String placeId) {
            SavePlaceDialog frag = new SavePlaceDialog();
            Bundle args = new Bundle();
            args.putString("title", title);
            args.putString("place_name", placeName);
            args.putString("place_id", placeId);
            frag.setArguments(args);
            return frag;
        }

        @Override
        @NonNull
        public Dialog onCreateDialog(Bundle savedInstanceState) {
            String title = getArguments().getString("title");
            final String placeName = getArguments().getString("place_name");
            final String placeId = getArguments().getString("place_id");
            AlertDialog.Builder alertDialogBuilder = new AlertDialog.Builder(getActivity());
            alertDialogBuilder.setTitle(title);
            alertDialogBuilder.setMessage(String.format(getString(R.string.save_this_place), placeName));
            alertDialogBuilder.setPositiveButton(getString(R.string.ok),  new DialogInterface.OnClickListener() {
                @Override
                public void onClick(DialogInterface dialog, int which) {
                    // on success
                    self.savePlace(placeId, placeName);
                    self.returnPlace(placeId, placeName);
                    dialog.dismiss();
                }
            });
            alertDialogBuilder.setNegativeButton(getString(R.string.cancel), new DialogInterface.OnClickListener() {
                @Override
                public void onClick(DialogInterface dialog, int which) {
                    self.returnPlace(placeId, placeName);
                    dialog.dismiss();
                }
            });

            return alertDialogBuilder.create();
        }
    }

}

File: app/src/main/java/com/xomena/cmpfutboltfe/SelectPlaceActivity.kt
package com.xomena.cmpfutboltfe

import com.xomena.cmpfutboltfe.model.*
import com.xomena.cmpfutboltfe.util.*
import com.xomena.cmpfutboltfe.ui.adapter.*

import android.app.Dialog
import android.content.Context
import android.content.Intent
import android.content.SharedPreferences
import android.graphics.drawable.Drawable
import android.os.Bundle
import android.util.Log
import android.view.View
import android.widget.AdapterView
import android.widget.AutoCompleteTextView
import android.widget.Toast

import androidx.annotation.NonNull
import androidx.appcompat.app.ActionBar
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.Toolbar
import androidx.core.content.res.ResourcesCompat
import androidx.fragment.app.DialogFragment
import androidx.fragment.app.FragmentManager
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView

import com.google.android.gms.common.ConnectionResult
import com.google.android.gms.common.api.GoogleApiClient
import com.google.android.gms.common.api.ResultCallback
import com.google.android.gms.location.places.AutocompletePrediction
import com.google.android.gms.location.places.Place
import com.google.android.gms.location.places.PlaceBuffer
import com.google.android.gms.location.places.Places

class SelectPlaceActivity : AppCompatActivity(),
    GoogleApiClient.OnConnectionFailedListener,
    SelectPlacesAdapter.OnItemClickListener {

    private lateinit var mGoogleApiClient: GoogleApiClient
    private lateinit var mAdapter: PlacesAutocompleteAdapter
    private lateinit var adapter: SelectPlacesAdapter

    private val mAutocompleteClickListener = AdapterView.OnItemClickListener { parent, view, position, id ->
        /*
         Retrieve the place ID of the selected item from the Adapter.
         The adapter stores each Place suggestion in a AutocompletePrediction from which we
         read the place ID and title.
          */
        val item = mAdapter.getItem(position)
        val placeId = item?.placeId
        val primaryText = item?.getPrimaryText(null)

        if (placeId != null && primaryText != null) {
            val fm = supportFragmentManager
            val alertDialog = SavePlaceDialog.newInstance(
                getString(R.string.title_activity_select_place),
                primaryText.toString(), placeId
            )
            alertDialog.show(fm, "fragment_alert")
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        self = this

        mGoogleApiClient = GoogleApiClient.Builder(this)
            .enableAutoManage(this, 0 /* clientId */, this)
            .addApi(Places.GEO_DATA_API)
            .build()

        setContentView(R.layout.activity_select_place)
        val toolbar = findViewById<Toolbar>(R.id.toolbarSelectPlace)
        setSupportActionBar(toolbar)
        try {
            val ab = supportActionBar
            ab?.setDisplayShowTitleEnabled(false)
        } catch (e: NullPointerException) {
            Log.e(LOG_TAG, "Exception", e)
        }

        val upArrow = ResourcesCompat.getDrawable(
            resources, R.drawable.arrow_left,
            applicationContext.theme
        )
        toolbar.navigationIcon = upArrow
        toolbar.setNavigationOnClickListener { onBackPressed() }

        val mAutocompleteView = findViewById<AutoCompleteTextView>(R.id.fieldFindPlace)
        mAutocompleteView.onItemClickListener = mAutocompleteClickListener
        mAdapter = PlacesAutocompleteAdapter(this, mGoogleApiClient, MainActivity.BOUNDS_TENERIFE, null)
        mAutocompleteView.setAdapter(mAdapter)

        val recyclerView = findViewById<RecyclerView>(R.id.rvSelectPlace)

        adapter = SelectPlacesAdapter(MyPlace.createPlacesList(this))
        adapter.setOnItemClickListener(this)

        recyclerView.adapter = adapter
        // Set layout manager to position the items
        recyclerView.layoutManager = LinearLayoutManager(this)
        val itemDecoration: RecyclerView.ItemDecoration =
            DividerItemDecoration(this, DividerItemDecoration.VERTICAL_LIST)
        recyclerView.addItemDecoration(itemDecoration)
    }

    override fun onConnectionFailed(@NonNull connectionResult: ConnectionResult) {
        Log.e(
            LOG_TAG, "onConnectionFailed: ConnectionResult.getErrorCode() = "
                    + connectionResult.errorCode
        )

        Toast.makeText(
            this,
            "Could not connect to Google API Client: Error " + connectionResult.errorCode,
            Toast.LENGTH_SHORT
        ).show()
    }

    override fun onItemClick(itemView: View, position: Int) {
        val place = adapter.getItem(position)
        if (place != null) {
            returnPlace(place.placeID, place.address)
        } else {
            val resultIntent = Intent()
            setResult(RESULT_CANCELED, resultIntent)
            finish()
        }
    }

    private fun savePlace(placeId: String, placeName: String) {
        val sharedPref = this.getSharedPreferences(
            MainActivity.PLACES_SHARED_PREF, Context.MODE_PRIVATE
        )

        val keys = sharedPref.getStringSet(STORED_KEYS, LinkedHashSet()) ?: LinkedHashSet()
        if (keys.contains(placeId)) {
            val toastMsg = String.format(getString(R.string.place_alredy_in_list), placeName)
            Toast.makeText(this, toastMsg, Toast.LENGTH_LONG).show()
        } else {
            Places.GeoDataApi.getPlaceById(mGoogleApiClient, placeId)
                .setResultCallback(object : ResultCallback<PlaceBuffer> {
                    override fun onResult(@NonNull places: PlaceBuffer) {
                        if (places.status.isSuccess) {
                            val place = places[0]

                            val placeData = LinkedHashSet<String>()
                            val val_str = place.id + "###" + place.latLng.latitude.toString() +
                                    "###" + place.latLng.longitude.toString() +
                                    "###" + place.name.toString() +
                                    "###" + place.address.toString()
                            placeData.add(val_str)

                            val newKeys = LinkedHashSet(keys)
                            newKeys.add(place.id)

                            val editor = sharedPref.edit()
                            editor.putStringSet(STORED_KEYS, newKeys)
                            editor.putStringSet(place.id, placeData)
                            editor.apply()

                            val toastMsg = String.format(getString(R.string.place_added_to_list), place.name)
                            Toast.makeText(self, toastMsg, Toast.LENGTH_LONG).show()
                        }
                        //Release the PlaceBuffer to prevent a memory leak
                        places.release()
                    }
                })
        }
    }

    private fun returnPlace(placeId: String?, placeAddress: String?) {
        val resultIntent = Intent()
        resultIntent.putExtra(MainActivity.EXTRA_PLACEID, placeId)
        resultIntent.putExtra(MainActivity.EXTRA_ADDRESS, placeAddress)
        setResult(RESULT_OK, resultIntent)
        finish()
    }

    class SavePlaceDialog : DialogFragment() {

        @NonNull
        override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
            val title = arguments?.getString("title")
            val placeName = arguments?.getString("place_name")
            val placeId = arguments?.getString("place_id")
            val alertDialogBuilder = AlertDialog.Builder(requireActivity())
            alertDialogBuilder.setTitle(title)
            alertDialogBuilder.setMessage(String.format(getString(R.string.save_this_place), placeName))
            alertDialogBuilder.setPositiveButton(getString(R.string.ok)) { dialog, _ ->
                // on success
                if (placeId != null && placeName != null) {
                    self?.savePlace(placeId, placeName)
                    self?.returnPlace(placeId, placeName)
                }
                dialog.dismiss()
            }
            alertDialogBuilder.setNegativeButton(getString(R.string.cancel)) { dialog, _ ->
                if (placeId != null && placeName != null) {
                    self?.returnPlace(placeId, placeName)
                }
                dialog.dismiss()
            }

            return alertDialogBuilder.create()
        }

        companion object {
            fun newInstance(title: String, placeName: String, placeId: String): SavePlaceDialog {
                val frag = SavePlaceDialog()
                val args = Bundle()
                args.putString("title", title)
                args.putString("place_name", placeName)
                args.putString("place_id", placeId)
                frag.arguments = args
                return frag
            }
        }
    }

    companion object {
        private var self: SelectPlaceActivity? = null
        private const val LOG_TAG = "SelectPlaceActivity"
        const val STORED_KEYS = "PLACE_IDS"
    }
}
